src/git/tree.mli

```diff
@@ -1,38 +1,78 @@
-(*
- * Copyright (c) 2013-2017 Thomas Gazagnaire <thomas@gazagnaire.org>
- *
- * Permission to use, copy, modify, and distribute this software for any
- * purpose with or without fee is hereby granted, provided that the above
- * copyright notice and this permission notice appear in all copies.
- *
- * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
- * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
- * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
- * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
- * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
- * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
- * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
- *)
-
-type perm =
-  [ `Normal
-  | `Exec
-  | `Link
-  | `Dir
-  | `Commit ]
-
-val pretty_perm: perm -> string
-val string_of_perm : perm -> string
-val fixed_length_string_of_perm : perm -> string
-
-type entry = {
-  perm: perm;
-  name: string;
-  node: Hash.t;
-}
-
-type t = entry list
-
-include S.S with type t := t
-
-module IO (D: Hash.DIGEST): S.IO with type t = t
+module type S =
+sig
+  module Digest
+    : Ihash.IDIGEST
+  (** The [Digest] module used to make the module. *)
+
+  module Hash
+    : Common.BASE
+  (** The Hash module *)
+
+  type entry =
+    { perm : perm
+    ; name : string
+    ; node : Hash.t }
+  and perm =
+    [ `Normal (** A {!Blob.t}. *)
+    | `Everybody
+    | `Exec   (** An executable. *)
+    | `Link   (** A {!Blob.t} that specifies the path of a {i symlink}. *)
+    | `Dir    (** A sub-{!Tree.t}. *)
+    | `Commit (** A sub-module ({!Commit.t}). *)
+    ]
+  and t = entry list
```

Comment:
The entry list is supposed to stay sorted (with a weird lexicographic order distinguishing tree and objects). Do you manage this? If yes, I guess the type `t` would be better be `and t = private entry list` to avoid random orders.